### PR TITLE
Flatten devices store

### PR DESCRIPTION
**Pull request by chrisnojima**
This PR explores some best practices for redux and immutable. I took the devices screen since its simple and self contained so I could convert it wholesale.

- [x] Uses entities for devices
- [x] Immutable for devices
- [x] Connect individual device rows
- [x] HOC for passing simple props to renderers
- [x] I explored not using action creators since they often don't do much, but, one important thing they *do* do is deal with hiding `payload: undefined` and the flow typing. Doing the flow typing inline is much grosser looking and easy to forget

Fixes a bug also
- [x] Don't regenerate paperkeys when you go back to devices tab

Note: I was iterating on this a bunch so i didn't make nice commits. It might be easier to look at this diff split, or ignore the old version

### CODE REVIEW

File: shared/actions/login/index.js [hunk 1]
```diff
@@ -329,10 +329,6 @@ export function addNewComputer () : AsyncAction {
   return addNewDevice(Constants.codePageDeviceRoleNewComputer)
 }
 
-export function addNewPaperKey () : Action {
```
Comment (chrisnojima) [inline shared/actions/login/index.js]: nothing outside of devices uses this

File: shared/constants/entities.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+// @flow
+import {Map, Record} from 'immutable'
+
+import type {NoErrorTypedAction} from './types/flux'
+import type {DeviceDetailRecord} from './devices'
+
+export type EntityType = any // TODO stronger typing?
```
Comment (chrisnojima) [inline shared/constants/entities.js]: Is there a way to make a map of the keypath to the types?

File: shared/devices/container.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import {connect} from 'react-redux'
+import {createSelector} from 'reselect'
+import {load, paperKeyMake} from '../actions/devices'
+
+import type {TypedState} from '../constants/reducer'
+
+const getAllDevicesSelector = (state: TypedState) => state.devices.get('deviceIDs')
+const getDeviceEntitiesSelector = (state: TypedState) => state.entities.get('devices')
+
+const getDevicesAndRevokedDevicesSelector = createSelector(
```
Comment (chrisnojima) [inline shared/devices/container.js]: Use selectors to avoid recalculating this

File: shared/devices/container.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    const deviceIDs = split.get('deviceIDs', List())
+    const revokedDeviceIDs = split.get('revokedDeviceIDs', List())
+    return {
+      deviceIDs,
+      revokedDeviceIDs,
+    }
+  }
+)
+
+const mapStateToProps = (state: any, {routeState}) => {
```
Comment (chrisnojima) [inline shared/devices/container.js]: seems a little cleaner to break these out into their own functions

File: shared/devices/container.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+const mapDispatchToProps = (dispatch: any, {routeState, setRouteState}) => ({
+  addNewComputer: () => dispatch(addNewComputer()),
+  addNewPaperKey: () => dispatch(paperKeyMake()),
+  addNewPhone: () => dispatch(addNewPhone()),
+  loadDevices: () => dispatch(load()),
+  onToggleShowRevoked: () => { setRouteState({showingRevoked: !routeState.showingRevoked}) },
+})
+
+const menuItems = props => ([
```
Comment (chrisnojima) [inline shared/devices/container.js]: Hoisted up some shared stuff

File: shared/devices/container.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+])
+
+const Devices = compose(
+  lifecycle({
+    componentWillMount: function () {
+      this.props.loadDevices()
+    },
+  }),
+  // Don't pass immutable things to dumb components
+  mapProps(props => ({
```
Comment (chrisnojima) [inline shared/devices/container.js]: I didn't touch the functionality of the render functions that much. Mostly hoisted up some props and reduced some code duplication. In general I think the dumb components should be a lot dumber than they are now. They should always take simple json props and get flat props that they consume simply. Dumb components knowing deeply about what a message is and types and things like that makes them unportable

File: shared/devices/device-page/container.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  'backup': 'paper key',
+  'desktop': 'device',
+  'mobile': 'device',
+}[props.device.type])
+
+const makeRenderProps = props => ({
+  ...props,
+  bannerBackgroundColor: bannerBackgroundColor(props),
+  bannerColor: bannerColor(props),
+  bannerDesc: null, // TODO at some point
```
Comment (chrisnojima) [inline shared/devices/device-page/container.js]: We haven't implemented this at all actually but its plumbed through

File: shared/devices/device-page/dumb.js [hunk 8]
```diff
@@ -1,91 +1,114 @@
 // @flow
-import Render from './index.render'
+import Render from '.'
```
Comment (chrisnojima) [inline shared/devices/device-page/dumb.js]: we could decide if we want to drive the dumb components directly or use the mock store and drive contained components. i see arguments for both sides. if we make things really dumb based we could use react storybook etc more effectively.

File: shared/devices/device-page/dumb.native.js [hunk 9]
```diff
@@ -1,92 +0,0 @@
-// @flow
```
Comment (chrisnojima) [inline shared/devices/device-page/dumb.native.js]: don't think we actually need a mobile specific version of most of these dumb sheets

File: shared/devices/device-revoke/container.js [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+import {connect} from 'react-redux'
+import {isMobile} from '../../constants/platform'
+import {navigateUp} from '../../actions/route-tree'
+import {revoke} from '../../actions/devices'
+
+import type {TypedState} from '../../constants/reducer'
+
+const mapStateToProps = (state: TypedState, {routeProps}) => ({
+  // $FlowIssue getIn
+  device: state.entities.getIn(['devices', routeProps.deviceID]),
```
Comment (chrisnojima) [inline shared/devices/device-revoke/container.js]: annoying that flow doesn't understand these helpers. maybe we can enhance the flow-typed?

File: shared/devices/dumb.js [hunk 11]
```diff
@@ -1,75 +1,93 @@
 // @flow
-import Devices from './render'
-import type {Device} from '../constants/types/more'
+import Devices from './container'
```
Comment (chrisnojima) [inline shared/devices/dumb.js]: This sheet needs to drive things w/ a mock store as it holds the rows which are connected themselves

**Comment by chrisnojima** [comment]
@keybase/react-hackers this is ready for review!

**Comment by chrisnojima** [comment]
Also please try the various flows. See devices, revoke keys, make a paperkey etc.

**Comment by chrisnojima** [comment]
@keybase/react-hackers bump

**Comment by keybase-ci-visdiff** [comment]
The commits b46ef740fe22...1c8db7952f68 introduce visual changes on linux. <details><summary>:mag_right: 5 changed</summary>

 * Changed: **Success-Start** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489442534552-b46ef740fe22-1c8db7952f68-linux/Success-Start-b46ef740fe22.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489442534552-b46ef740fe22-1c8db7952f68-linux/Success-Start-1c8db7952f68.png)  
   Success-Start comparison
 * Changed: **Devices_List-Devices** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489442534552-b46ef740fe22-1c8db7952f68-linux/Devices_List-Devices-b46ef740fe22.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489442534552-b46ef740fe22-1c8db7952f68-linux/Devices_List-Devices-1c8db7952f68.png)  
   Devices_List-Devices comparison
 * Changed: **Device_Page-Unlock** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489442534552-b46ef740fe22-1c8db7952f68-linux/Device_Page-Unlock-b46ef740fe22.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489442534552-b46ef740fe22-1c8db7952f68-linux/Device_Page-Unlock-1c8db7952f68.png)  
   Device_Page-Unlock comparison
 * Changed: **Device_Page-Paper** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489442534552-b46ef740fe22-1c8db7952f68-linux/Device_Page-Paper-b46ef740fe22.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489442534552-b46ef740fe22-1c8db7952f68-linux/Device_Page-Paper-1c8db7952f68.png)  
   Device_Page-Paper comparison
 * Changed: **Device_Page-Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489442534552-b46ef740fe22-1c8db7952f68-linux/Device_Page-Normal-b46ef740fe22.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489442534552-b46ef740fe22-1c8db7952f68-linux/Device_Page-Normal-1c8db7952f68.png)  
   Device_Page-Normal comparison
</summary>

File: shared/login/signup/success/index.render.desktop.js [hunk 12]
```diff
@@ -19,14 +19,24 @@ class SuccessRender extends Component<void, Props, State> {
...
 
   render () {
-    return (
-      <Container onBack={this.props.onBack} style={stylesContainer}>
-        <Text type='Header' style={stylesHeader}>{this.props.title || "Congratulations, you've just joined Keybase!"}</Text>
-        <Text type='Body' style={stylesBody}>Here is your unique paper key, it will allow you to perform important Keybase tasks in the future. This is the only time you'll see this so be sure to write it down.</Text>
+    const contents = this.props.paperkey
+      ? (
         <Box style={stylesPaperKeyContainer}>
           <Text type='Header' style={stylesPaperkey}>{this.props.paperkey.stringValue()}</Text>
```
Comment (chromakode) [inline shared/login/signup/success/index.render.desktop.js]: Need `<Icon type='icon-paper-key-corner' style={stylesPaperCorner} />` here too

File: shared/devices/device-page/index.desktop.js [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    </Box>
+    {!!bannerDesc && <Banner color={bannerColor} backgroundColor={bannerBackgroundColor} desc={bannerDesc} />}
+    <Box style={{...globalStyles.flexBoxRow, padding: 30}}>
+      <Box style={{...globalStyles.flexBoxRow, alignItems: 'flex-start', justifyContent: 'center', width: 240}}>
+        <Icon type={icon} style={{opacity: revokedAt ? 0.4 : 1}} />
+      </Box>
+      <Box style={{...globalStyles.flexBoxColumn}}>
+        <Header name={name} currentDevice={currentDevice} revokedAt={revokedAt} />
+        {!!timeline && <Timeline timeline={timeline} />}
+        {!revokedAt && <Button type='Danger' style={{marginTop: 15}} label={`Revoke this ${revokeName || ''}`} onClick={showRevokeDevicePage} />}
```
Comment (chromakode) [inline shared/devices/device-page/index.desktop.js]: The revoke name seems to be missing in the latest visdiff run. Not getting plumbed through to here?

**Comment by chrisnojima** [comment]
- [ ] fix visdiff

File: shared/devices/device-page/index.desktop.js (hunk 13 above)
Comment (chrisnojima) [inline shared/devices/device-page/index.desktop.js]: this is cause it got moved around and the dumbsheet had it as null

File: shared/login/signup/success/index.render.desktop.js (hunk 12 above)
Comment (chrisnojima) [inline shared/login/signup/success/index.render.desktop.js]: tnx!

File: shared/devices/index.js.flow [hunk 14]
```diff
@@ -0,0 +1,20 @@
+// @flow
+import {Component} from 'react'
+
+import type {Device} from '../constants/types/flow-types'
+import type {MenuItem} from '../common-adapters/popup-menu.js'
+
+export type Props = {
+  deviceIDs: Array<string>,
+  // $FlowIssue not sure why menuitem conflicts
+  menuItems: Array<MenuItem>,
```
Comment (chromakode) [inline shared/devices/index.js.flow]: Making this `MenuItem | 'Divider' | null` will fix the flow issue. Flow can't generalize `Array<MenuItem>` => `Array<MenuItem | 'Divider' | null>`, which is what the PopupMenu component expects.

File: shared/actions/devices.js [hunk 15]
```diff
@@ -1,130 +1,142 @@
...
+        name: d.device.name,
+        provisionedAt: d.provisionedAt,
+        revokedAt: d.revokedAt,
+        revokedBy: d.revokedByDevice,
+        type: d.device.type,
+      })
+      return map
+    }, {})
+
+    const deviceIDs = result
```
Comment (chrisnojima) [inline shared/actions/devices.js]: i changed this around and i think its clearer now, tnx

File: shared/constants/entities.js [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+export type Merge = NoErrorTypedAction<'entity:merge', {keyPath: Array<string>, entities: {[id: string]: EntityType}}>
+export type Replace = NoErrorTypedAction<'entity:replace', {keyPath: Array<string>, entities: {[id: string]: EntityType}}>
+
+export type Actions = Delete
+  | Merge
+  | Replace
+
+// State
+export type State = Record<{
+  devices: Map<string, DeviceDetailRecord>,
```
Comment (chrisnojima) [inline shared/constants/entities.js]: i figured all entities would always be keyed by strings. seems simplest right?

File: shared/devices/container.js [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+import {createSelector} from 'reselect'
+import {load, paperKeyMake} from '../actions/devices'
+
+import type {TypedState} from '../constants/reducer'
+
+const getAllDevicesSelector = (state: TypedState) => state.devices.get('deviceIDs')
+const getDeviceEntitiesSelector = (state: TypedState) => state.entities.get('devices')
+
+const getDevicesAndRevokedDevicesSelector = createSelector(
+  getAllDevicesSelector, getDeviceEntitiesSelector,
```
Comment (chrisnojima) [inline shared/devices/container.js]: sure

File: shared/actions/devices.js [hunk 18]
```diff
@@ -1,130 +1,142 @@
...
+      .sort(_sortDevices)
+      .map(d => d.device.deviceID)
+
+    yield put(replaceEntity(['devices'], entities))
+    yield put(loaded(deviceIDs))
+    yield put(setWaiting(false))
```
Comment (chrisnojima) [inline shared/actions/devices.js]: nope, was a leftover before i had all the finally clauses :)

File: shared/devices/index.js.flow (hunk 14 above)
Comment (chrisnojima) [inline shared/devices/index.js.flow]: was hoping that wasn't the answer , oh well

**Comment by chrisnojima** [comment]
@chromakode @MarcoPolo updated with your feedback, please tal

**Comment by keybase-ci-visdiff** [comment]
The commits b46ef740fe22...23951ddc3c44 introduce visual changes on linux. <details><summary>:mag_right: 3 changed</summary>

 * Changed: **Success-Start** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489497362285-b46ef740fe22-23951ddc3c44-linux/Success-Start-b46ef740fe22.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489497362285-b46ef740fe22-23951ddc3c44-linux/Success-Start-23951ddc3c44.png)  
   Success-Start comparison
 * Changed: **Devices_List-Devices** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489497362285-b46ef740fe22-23951ddc3c44-linux/Devices_List-Devices-b46ef740fe22.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489497362285-b46ef740fe22-23951ddc3c44-linux/Devices_List-Devices-23951ddc3c44.png)  
   Devices_List-Devices comparison
 * Changed: **Device_Page-Paper** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489497362285-b46ef740fe22-23951ddc3c44-linux/Device_Page-Paper-b46ef740fe22.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489497362285-b46ef740fe22-23951ddc3c44-linux/Device_Page-Paper-23951ddc3c44.png)  
   Device_Page-Paper comparison
</summary>

**Comment by chrisnojima** [comment]
re: visdiff. I'll fix the button text. but the device list showing revoked i think is likely better. i put a min height for the paperkey so it can show a progress and allow a null paperkey since this is a re-used screen (instead of crashing out). I'll center the paperkey so it doesn't look offset

**Comment by keybase-ci-visdiff** [comment]
The commits b46ef740fe22...7a11c8f400db introduce visual changes on linux. <details><summary>:mag_right: 2 changed</summary>

 * Changed: **Success-Start** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489499082838-b46ef740fe22-7a11c8f400db-linux/Success-Start-b46ef740fe22.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489499082838-b46ef740fe22-7a11c8f400db-linux/Success-Start-7a11c8f400db.png)  
   Success-Start comparison
 * Changed: **Devices_List-Devices** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489499082838-b46ef740fe22-7a11c8f400db-linux/Devices_List-Devices-b46ef740fe22.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1489499082838-b46ef740fe22-7a11c8f400db-linux/Devices_List-Devices-7a11c8f400db.png)  
   Devices_List-Devices comparison
</summary>

**Pull request merged**